crates/proto/src/clients/mod.rs

```diff
@@ -0,0 +1,370 @@
+        let mut metadata = MetadataInterceptor::default();
+        if let Some(version) = config.metadata_version.as_deref() {
+            metadata = metadata
+                .with_accept_metadata(version, config.metadata_genesis.as_deref())
+                .expect("Failed to create metadata interceptor");
+        }
+        let combined = OtelAndMetadataInterceptor::new(OtelInterceptor, metadata);
+        generated::rpc::api_client::ApiClient::with_interceptor(channel, combined)
+    }
+}
+
+impl GrpcClientBuilder for BlockProducer {
+    type Service = BlockProducerClient;
+
+    fn with_interceptor(channel: Channel, _config: &ClientConfig) -> Self::Service {
+        generated::block_producer::api_client::ApiClient::with_interceptor(channel, OtelInterceptor)
+    }
+}
+
+impl GrpcClientBuilder for StoreNtxBuilder {
+    type Service = StoreNtxBuilderClient;
+
+    fn with_interceptor(channel: Channel, _config: &ClientConfig) -> Self::Service {
+        generated::ntx_builder_store::ntx_builder_client::NtxBuilderClient::with_interceptor(
+            channel,
+            OtelInterceptor,
+        )
+    }
+}
+
+impl GrpcClientBuilder for StoreBlockProducer {
+    type Service = StoreBlockProducerClient;
+
+    fn with_interceptor(channel: Channel, _config: &ClientConfig) -> Self::Service {
+        generated::block_producer_store::block_producer_client::BlockProducerClient::with_interceptor(
+            channel,
+            OtelInterceptor,
+        )
+    }
+}
+
+impl GrpcClientBuilder for StoreRpc {
+    type Service = StoreRpcClient;
+
+    fn with_interceptor(channel: Channel, _config: &ClientConfig) -> Self::Service {
+        generated::rpc_store::rpc_client::RpcClient::with_interceptor(channel, OtelInterceptor)
+    }
+}
+
+// STRICT TYPE-SAFE BUILDER (NO DEFAULTS)
+// ================================================================================================
+
+/// A type-safe builder that forces the caller to make an explicit decision for each
+/// configuration item (TLS, timeout, metadata version, metadata genesis) before connecting.
+///
+/// This builder replaces the previous defaulted builder. Callers must explicitly choose TLS,
+/// timeout, and metadata options before connecting.
+///
+/// Usage example:
+///
+/// ```rust,no_run
+/// use miden_node_proto::clients::{Builder, WantsTls, Rpc, RpcClient};
+/// use std::time::Duration;
+///
+/// # async fn example() -> anyhow::Result<()> {
+/// let client: RpcClient = Builder::new("https://rpc.example.com:8080")?
+///     .with_tls()?                        // or `.without_tls()`
+///     .with_timeout(Duration::from_secs(5)) // or `.without_timeout()`
+///     .with_metadata_version("1.0".into()) // or `.without_metadata_version()`
+///     .without_metadata_genesis()           // or `.with_metadata_genesis(genesis)`
+///     .connect::<Rpc>()
+///     .await?;
+/// # Ok(())
+/// # }
+/// ```
+#[derive(Clone, Debug)]
+pub struct Builder<State> {
```

Comment:
note: we imply an ordering here, and in the associated functions, using a generic per assoticated function with a corresponding `Has*` to its `Wants*`, one could make the order arbitrary, which is what I'd expect as user. It comes with a compile time cost of extra generics and a bag of generated types. It's a trade-off.